### ISSUE TITLE
Avoid spawning a new thread for every call.

### ISSUE THREAD

**Issue by jonhoo** [google/tarpc#56]
Unless I'm mistaken, `tarpc` currently spawns a new thread [for every RPC](https://github.com/google/tarpc/blob/master/src/macros.rs#L573). For a high-traffic service where each RPC handler exist quickly, this adds a lot of overhead. It'd be nice if a thread pool could be used instead to save some of the overhead. Is this something that's on the roadmap? I noticed it was mentioned as a drawback in @shaladdle's Reddit post [announcing the project](https://www.reddit.com/r/rust/comments/46bzyz/announcing_tarpc_a_simple_rpc_library/d03wcun) in March, but haven't seen it mentioned anywhere else since.

**Comment by shaladdle** [comment]
Hey Jon, thanks for filing an issue. This is something we want to do, but it's a bit low priority at the moment.

We're in a transition mode as we move Tarpc to work on top of Tokio. The latest code is at tikue/framed.

After that branch is merged to master (and is published on crates.io), there will be two traits generated by the `service!` macro: `SyncService` and `FutureService`. The lowest overhead approach to writing a service using Tarpc is to implement the `FutureService` trait, and manage any threads you need yourself.

Once framed is merged I want to take a stab at this. For now hopefully implementing `FutureService` works for you.

**Comment by jonhoo** [comment]
Ah, I see. Thanks for the quick reply. I might give that a shot then. Is there an example for how to implement the `FutureService` trait somewhere?

**Comment by tikue** [comment]
The main thing with `FutureService` is to avoid blocking by always working within the [`Future`](http://alexcrichton.com/futures-rs/futures/index.html) abstraction.

Regarding a thread pool for `SyncService` -- one of the main issues is that there's no good default for a fixed size thread pool; it's very system-specific. Preferably we'd use a dynamically resizing thread pool that caches threads for a some reasonable amount of time (on the order of minutes?) and then expires them after an idle period. I [wrote something](https://github.com/google/tarpc/blob/e04f48e44486ef912422570398eec95264cb8519/src/cached_pool/mod.rs) like this a while ago but removed it because it had some issues and I didn't have the bandwidth to work on it. Maybe we'll resurrect it at some point.

**Comment by shaladdle** [comment]
Unfortunately, there's an ICE in nightly right now which breaks us: https://github.com/rust-lang/rust/issues/36943. Until that's fixed, you'll have to pin yourself to this nightly which doesn't have that bug:

```
rustup override set nightly-2016-09-22
```

I've sent tikue/tarpc/pull/4 to add a future example to the readme. It tried it and it works with that nightly.

**Comment by jonhoo** [comment]
Thanks for the example. I assume then that the `FutureService` methods are called synchronously from some kind of event loop with no multithreading, otherwise adding a thread pool there wouldn't actually make things any faster?

With regards to the issue of thread pool sizes, I think this could sensibly be a parameter when creating the service. Dynamic sizing is nice, but (as you observe) tricky to get right.

@shaladdle yeah, I've run into that ICE, and also an issue in the `aster` dependency of `serde` which prevents building on the currently nightly, so I've already overridden it.

**Comment by tikue** [comment]
@jonhoo yep that's correct, `FutureService` is based on top of [tokio](https://github.com/tokio-rs/), which is an async framework built on top of mio and futures-rs. All method impls in `FutureService` are expected to return very quickly, or else they'd be blocking the tokio reactor core. If everything is built on tarpc or tokio, you get this by default. Some things, like cpu-intensive work or file I/O, will need to use thread pools so as to not block the reactor core.

**Comment by jonhoo** [comment]
@tikue great, thanks, I'll give it a shot!

**Comment by jonhoo** [comment]
Hmm, when trying to build `framed` with the nightly @shaladdle gave, I see:
```console
$ cargo build
   Compiling quasi_macros v0.20.0
error: function `register` is private
  --> /home/jon/.cargo/registry/src/github.com-1ecc6299db9ec823/quasi_macros-0.20.0/src/lib.rs:21:5
   |
21 |     quasi_codegen::register(reg);
   |     ^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> /home/jon/.cargo/registry/src/github.com-1ecc6299db9ec823/quasi_macros-0.20.0/src/lib.rs:21:29
   |
21 |     quasi_codegen::register(reg);
   |                             ^^^ expected struct `syntex::Registry`, found struct `rustc_plugin::Registry`
   |
   = note: expected type `&mut syntex::Registry`
   = note:    found type `&mut rustc_plugin::Registry<'_>`

error: aborting due to previous error

error: Could not compile `quasi_macros`.
```

**Comment by jonhoo** [comment]
Okay, so I've written some code that I believe should be right using `FutureService`, but am having some issues. My code is something along the lines of:
```rust
impl FutureService for Server {
    type HelloFUt = futures::Oneshot<()>;

    fn hello(&self) -> Self::HelloFut {
        let (c, p) = futures::oneshot();
        // self.workers: threadpool::ThreadPool
        self.workers.execute(move || {
            c.complete(());
        });
        p
    }
}
```

However, this fails with a couple of errors. First:
```
error[E0271]: type mismatch resolving `<futures::Oneshot<()> as futures::Future>::Error == tarpc::util::Never`
  --> src/srv/mod.rs:56:6
   |
56 | impl FutureService for Server {
   |      ^^^^^^^^^^^^^ expected struct `futures::Canceled`, found struct `tarpc::util::Never`
   |
   = note: expected type `futures::Canceled`
   = note:    found type `tarpc::util::Never`
   = note: required by `srv::ext::FutureService`
```
I cannot control the error type of `OneShot` (it is always `Canceled`), but apparently it is required that it's set to `Never`. Why is this requirement there? Alternatively, is there another way for me to do a oneshot in which I can set the error type?

Second:
```
error[E0277]: the trait bound `srv::Server: std::clone::Clone` is not satisfied
  --> src/srv/mod.rs:56:6
   |
56 | impl FutureService for Server {
   |      ^^^^^^^^^^^^^ trait `srv::Server: std::clone::Clone` not satisfied
   |
   = note: required by `srv::ext::FutureService`
```
Why is it necessary that my server struct is `Clone`? That seems like a peculiar requirement? Hopefully there is only one instance of my `Server` running?

**Comment by jonhoo** [comment]
Digging through the code, I just found
```rust
/// Defines the `Future` RPC service. Implementors must be `Clone`, `Send`, and `'static`,
/// as required by `tokio_proto::NewService`. This is required so that the service can be used
/// to respond to multiple requests concurrently.
```
This should probably be put somewhere in the `tarpc` documentation, as the docstring from inside the macro will only be visible in the doc generated for the user's code. I'm still curious as to why the `tokio_proto` requires `Clone` and not just `Sync`, but I guess that's a question for them, not you.

**Comment by jonhoo** [comment]
The `Future::Error` type issue seems tricky too. Only by exploring the macro code did I find the `| error` syntax, which apparently defaults to `::util::Never`. Changing my functions to all ahve `| futures::Canceled` seems to have satisfied the compiler for now.

**Comment by jonhoo** [comment]
Looking some more at the `NewService` `Clone` requirement, this seems to be specific to [`NewService`](https://tokio-rs.github.io/tokio-service/tokio_service/trait.NewService.html). The [`Service` trait](https://tokio-rs.github.io/tokio-service/tokio_service/trait.Service.html) does not have this restriction. As far as I can tell, only the latter is required by the server creation mechanisms in [`tokio_proto`](https://tokio-rs.github.io/tokio-proto/tokio_proto/index.html)? If this is indeed all that is used, relaxing this restriction would be great.

**Comment by jonhoo** [comment]
Implemented in https://github.com/tikue/tarpc/pull/5.

**Comment by jonhoo** [comment]
As I suspected my little `| futures::Canceled` trick didn't actually work:
```
error[E0277]: the trait bound `futures::Canceled: query::_IMPL_SERIALIZE_FOR_DataType::_serde::Deserialize` is not satisfied
  --> src/srv/mod.rs:26:5
   |
26 |     service! {
   |     ^ trait `futures::Canceled: query::_IMPL_SERIALIZE_FOR_DataType::_serde::Deserialize` not satisfied
   |
   = note: this error originates in a macro from the standard library
```
Is there some other way for me to use a `futures::OneShot` inside a handler?

**Comment by tikue** [comment]
The error type is whatever error you intend to forward to the client. Since the `Canceled` error is kind of implementation specific (e.g. if you stop using `Oneshot` you'd need to modify the public API to not have Error = `Canceled`), you probably don't want to expose it directly. The typical way to do this is to have some kind of `InternalError` type that has like, a `String` description. Because this is so common, tarpc provides `tarpc::util::Message` which just wraps a string. This is done in the pubsub [example](https://github.com/google/tarpc/blob/master/examples/pubsub.rs#L109).

**Comment by tikue** [comment]
By the way, it may be quicker to ask these questions on our gitter channel, https://gitter.im/tarpc/Lobby.

**Comment by tikue** [comment]
This is fixed in master now with `FutureService`, and there will be a crates.io release probably within two weeks.

**Comment by jonhoo** [comment]
@tikue any update on a new crates.io release?

**Comment by tikue** [comment]
Sorry for the delay! It should come this week. I can't promise it, but I'd be surprised if we don't manage to ship it within a few days.

**Comment by jonhoo** [comment]
Maybe I'm misunderstanding, but the tokio-based `FutureService` seems to be limited to *one* thread, which is the entirely other extreme? Or rather, if you want a multi-threaded RPC server, you would need to `.listen()` multiple times with different reactors, and then drive them separately?

**Comment by tikue** [comment]
Yes, for now you need to spin up more reactors manually and initialize the server options each with a different handle. I expect the parallelism story to improve as tokio matures.

**Comment by jonhoo** [comment]
Actually, on second though, even that seems tricky. The way `.listen()` currently works, it will bind to a port, so if you want multiple threads that use multiple reactors, they will have to listen to *different* ports, right?

**Comment by tikue** [comment]
Tarpc configures listeners to allow the reuse of ports, so it should work. Feel free to open an issue if you run into trouble with it.
